### ISSUE TITLE
Potential deadlock during JWT credentials rotation

### ISSUE THREAD

**Issue by DaveCTurner** [elastic/elasticsearch#99317]
We observed a node (running a somewhat nonstandard build of ES in a fairly unusual environment) which was deadlocked while invalidating a JWT credentials cache entry while rotation credentials. It looked like it might be possible to experience the same deadlock in regular ES too.

```
Found one Java-level deadlock:
=============================
"elasticsearch[es-es-search-5577dfb9fd-fcjtw][transport_worker][T#12]":
  waiting for ownable synchronizer 0x000000070826d178, (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync),
  which is held by "elasticsearch[es-es-search-5577dfb9fd-fcjtw][transport_worker][T#20]"

"elasticsearch[es-es-search-5577dfb9fd-fcjtw][transport_worker][T#20]":
  waiting for ownable synchronizer 0x0000000708ff86b8, (a java.util.concurrent.locks.ReentrantLock$NonfairSync),
  which is held by "elasticsearch[es-es-search-5577dfb9fd-fcjtw][system_critical_read][T#1]"

"elasticsearch[es-es-search-5577dfb9fd-fcjtw][system_critical_read][T#1]":
  waiting for ownable synchronizer 0x000000070826d178, (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync),
  which is held by "elasticsearch[es-es-search-5577dfb9fd-fcjtw][transport_worker][T#20]"
```

<details>
<summary>Expand for more complete stack traces</summary>

```
"elasticsearch[es-es-search-5577dfb9fd-fcjtw][transport_worker][T#12]":
    at jdk.internal.misc.Unsafe.park(java.base@20.0.2/Native Method)
    - parking to wait for  <0x000000070826d178> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
    at java.util.concurrent.locks.LockSupport.park(java.base@20.0.2/LockSupport.java:221)
    at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(java.base@20.0.2/AbstractQueuedSynchronizer.java:754)
    at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireShared(java.base@20.0.2/AbstractQueuedSynchronizer.java:1079)
    at java.util.concurrent.locks.ReentrantReadWriteLock$ReadLock.lock(java.base@20.0.2/ReentrantReadWriteLock.java:738)
    at org.elasticsearch.common.util.concurrent.ReleasableLock.acquire(org.elasticsearch.server@8.11.0-SNAPSHOT/ReleasableLock.java:43)
    at org.elasticsearch.common.cache.Cache$CacheSegment.get(org.elasticsearch.server@8.11.0-SNAPSHOT/Cache.java:199)
    at org.elasticsearch.common.cache.Cache.get(org.elasticsearch.server@8.11.0-SNAPSHOT/Cache.java:350)
    at org.elasticsearch.common.cache.Cache.get(org.elasticsearch.server@8.11.0-SNAPSHOT/Cache.java:345)
    at org.elasticsearch.xpack.security.authc.jwt.JwtRealm.tryAuthenticateWithCache(org.elasticsearch.security@8.11.0-SNAPSHOT/JwtRealm.java:348)


"elasticsearch[es-es-search-5577dfb9fd-fcjtw][transport_worker][T#20]":
    at jdk.internal.misc.Unsafe.park(java.base@20.0.2/Native Method)
    - parking to wait for  <0x0000000708ff86b8> (a java.util.concurrent.locks.ReentrantLock$NonfairSync)
    at java.util.concurrent.locks.LockSupport.park(java.base@20.0.2/LockSupport.java:221)
    at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(java.base@20.0.2/AbstractQueuedSynchronizer.java:754)
    at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(java.base@20.0.2/AbstractQueuedSynchronizer.java:990)
    at java.util.concurrent.locks.ReentrantLock$Sync.lock(java.base@20.0.2/ReentrantLock.java:153)
    at java.util.concurrent.locks.ReentrantLock.lock(java.base@20.0.2/ReentrantLock.java:322)
    at org.elasticsearch.common.util.concurrent.ReleasableLock.acquire(org.elasticsearch.server@8.11.0-SNAPSHOT/ReleasableLock.java:43)
    at org.elasticsearch.common.cache.Cache.invalidateAll(org.elasticsearch.server@8.11.0-SNAPSHOT/Cache.java:529)
    at org.elasticsearch.xpack.security.authc.jwt.JwtRealm.invalidateJwtCache(org.elasticsearch.security@8.11.0-SNAPSHOT/JwtRealm.java:453)
    at org.elasticsearch.xpack.security.authc.jwt.JwtRealm.expireAll(org.elasticsearch.security@8.11.0-SNAPSHOT/JwtRealm.java:181)
    at org.elasticsearch.xpack.security.authc.jwt.JwtRealm$$Lambda$4053/0x0000000101c181f0.reloaded(org.elasticsearch.security@8.11.0-SNAPSHOT/Unknown Source)
    at org.elasticsearch.xpack.security.authc.jwt.JwtSignatureValidator$PkcJwtSignatureValidator.lambda$validate$0(org.elasticsearch.security@8.11.0-SNAPSHOT/JwtSignatureValidator.java:311)
    at org.elasticsearch.xpack.security.authc.jwt.JwtSignatureValidator$PkcJwtSignatureValidator$$Lambda$8830/0x00000001025cdb48.accept(org.elasticsearch.security@8.11.0-SNAPSHOT/Unknown Source)
    at org.elasticsearch.action.ActionListener$2.onResponse(org.elasticsearch.server@8.11.0-SNAPSHOT/ActionListener.java:177)
    at org.elasticsearch.action.support.SubscribableListener$SuccessResult.complete(org.elasticsearch.server@8.11.0-SNAPSHOT/SubscribableListener.java:273)
    at org.elasticsearch.action.support.SubscribableListener.tryComplete(org.elasticsearch.server@8.11.0-SNAPSHOT/SubscribableListener.java:193)
    at org.elasticsearch.action.support.SubscribableListener.addListener(org.elasticsearch.server@8.11.0-SNAPSHOT/SubscribableListener.java:96)
    at org.elasticsearch.action.support.SubscribableListener.addListener(org.elasticsearch.server@8.11.0-SNAPSHOT/SubscribableListener.java:71)
    at org.elasticsearch.xpack.security.authc.jwt.JwkSetLoader.reload(org.elasticsearch.security@8.11.0-SNAPSHOT/JwkSetLoader.java:90)
    at org.elasticsearch.xpack.security.authc.jwt.JwtSignatureValidator$PkcJwtSignatureValidator.validate(org.elasticsearch.security@8.11.0-SNAPSHOT/JwtSignatureValidator.java:292)
    at org.elasticsearch.xpack.security.authc.jwt.JwtSignatureValidator$DelegatingJwtSignatureValidator.validate(org.elasticsearch.security@8.11.0-SNAPSHOT/JwtSignatureValidator.java:144)
    at org.elasticsearch.xpack.security.authc.jwt.JwtAuthenticator.validateSignature(org.elasticsearch.security@8.11.0-SNAPSHOT/JwtAuthenticator.java:101)
    at org.elasticsearch.xpack.security.authc.jwt.JwtAuthenticator.authenticate(org.elasticsearch.security@8.11.0-SNAPSHOT/JwtAuthenticator.java:93)
    at org.elasticsearch.xpack.security.authc.jwt.JwtRealm.authenticate(org.elasticsearch.security@8.11.0-SNAPSHOT/JwtRealm.java:291)
    at org.elasticsearch.xpack.security.authc.RealmsAuthenticator.lambda$consumeToken$4(org.elasticsearch.security@8.11.0-SNAPSHOT/RealmsAuthenticator.java:147)

"elasticsearch[es-es-search-5577dfb9fd-fcjtw][system_critical_read][T#1]":
    at jdk.internal.misc.Unsafe.park(java.base@20.0.2/Native Method)
    - parking to wait for  <0x000000070826d178> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
    at java.util.concurrent.locks.LockSupport.park(java.base@20.0.2/LockSupport.java:221)
    at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(java.base@20.0.2/AbstractQueuedSynchronizer.java:754)
    at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(java.base@20.0.2/AbstractQueuedSynchronizer.java:990)
    at java.util.concurrent.locks.ReentrantReadWriteLock$WriteLock.lock(java.base@20.0.2/ReentrantReadWriteLock.java:959)
    at org.elasticsearch.common.util.concurrent.ReleasableLock.acquire(org.elasticsearch.server@8.11.0-SNAPSHOT/ReleasableLock.java:43)
    at org.elasticsearch.common.cache.Cache$CacheSegment.remove(org.elasticsearch.server@8.11.0-SNAPSHOT/Cache.java:293)
    at org.elasticsearch.common.cache.Cache.evictEntry(org.elasticsearch.server@8.11.0-SNAPSHOT/Cache.java:764)
    at org.elasticsearch.common.cache.Cache.evict(org.elasticsearch.server@8.11.0-SNAPSHOT/Cache.java:755)
    at org.elasticsearch.common.cache.Cache.promote(org.elasticsearch.server@8.11.0-SNAPSHOT/Cache.java:746)
    at org.elasticsearch.common.cache.Cache.put(org.elasticsearch.server@8.11.0-SNAPSHOT/Cache.java:470)
    at org.elasticsearch.common.cache.Cache.put(org.elasticsearch.server@8.11.0-SNAPSHOT/Cache.java:457)
    at org.elasticsearch.xpack.security.authc.jwt.JwtRealm.lambda$processValidatedJwt$4(org.elasticsearch.security@8.11.0-SNAPSHOT/JwtRealm.java:408)
    at org.elasticsearch.xpack.security.authc.jwt.JwtRealm$$Lambda$7938/0x000000010251aaa8.accept(org.elasticsearch.security@8.11.0-SNAPSHOT/Unknown Source)
    at org.elasticsearch.action.ActionListener$2.onResponse(org.elasticsearch.server@8.11.0-SNAPSHOT/ActionListener.java:177)
    at org.elasticsearch.xpack.security.authc.jwt.JwtRealm.lambda$processValidatedJwt$5(org.elasticsearch.security@8.11.0-SNAPSHOT/JwtRealm.java:432)
    at org.elasticsearch.xpack.security.authc.jwt.JwtRealm$$Lambda$7957/0x000000010251cfa0.accept(org.elasticsearch.security@8.11.0-SNAPSHOT/Unknown Source)
    at org.elasticsearch.action.ActionListener$2.onResponse(org.elasticsearch.server@8.11.0-SNAPSHOT/ActionListener.java:177)
    at org.elasticsearch.xpack.security.authc.support.mapper.NativeRoleMappingStore.lambda$resolveRoles$18(org.elasticsearch.security@8.11.0-SNAPSHOT/NativeRoleMappingStore.java:381)
    at org.elasticsearch.xpack.security.authc.support.mapper.NativeRoleMappingStore$$Lambda$7969/0x000000010251e920.accept(org.elasticsearch.security@8.11.0-SNAPSHOT/Unknown Source)
    at org.elasticsearch.action.ActionListener$2.onResponse(org.elasticsearch.server@8.11.0-SNAPSHOT/ActionListener.java:177)
    at org.elasticsearch.xpack.security.authc.support.mapper.NativeRoleMappingStore.lambda$loadMappings$0(org.elasticsearch.security@8.11.0-SNAPSHOT/NativeRoleMappingStore.java:140)
    at org.elasticsearch.xpack.security.authc.support.mapper.NativeRoleMappingStore$$Lambda$7989/0x0000000102521370.accept(org.elasticsearch.security@8.11.0-SNAPSHOT/Unknown Source)
    at org.elasticsearch.action.ActionListener$2.onResponse(org.elasticsearch.server@8.11.0-SNAPSHOT/ActionListener.java:177)
    at org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(org.elasticsearch.server@8.11.0-SNAPSHOT/ContextPreservingActionListener.java:32)
    at org.elasticsearch.xpack.core.security.ScrollHelper$1.onResponse(org.elasticsearch.xcore@8.11.0-SNAPSHOT/ScrollHelper.java:97)
```

</details>

---

### Symptoms

In this case the node left the cluster and could not rejoin, reporting timeouts. In other cases it might remain a member of the cluster, but some requests to the node will hang.

To diagnose the problem, use `jstack` to take a stack dump of the problematic node, look for reports of a deadlock, and compare the stack traces to the ones listed above.

---

### Workaround

Disable caching for all JWT-type realms:

```
xpack.security.authc.realms.jwt.<realmname>.jwt.cache.ttl: "0m"
```

or 

```
xpack.security.authc.realms.jwt.<realmname>.jwt.cache.size: 0
```

Note that this may have a performance impact.

**Comment by jakelandis** [comment]
There are actually 2 bugs here. The first bug is a specific to the JWT realm (easily fixable) and the second bug is specific to our Cache implementation (possibly difficult to fix).  If we fix the first bug, this specific issue can be resolved but still leaves the door open for other deadlock scenerios. I will log a seperate issue for the Cache implementation bug. 

## JWT realm bug

The JWT realm supports rotation of the keys that can validate the signature of the JWT used for authentication. The general strategy for rotation is upon failure to validate the signature, then the key material (JWKs) will be reloaded automatically, if there are any changes to the key material then the cache is evicted and the signature validation will be attempted again. Of note, the reload is only only holds up the second authentication attempt (blocking for file based reloads, async call back for http) on the single thread that triggered the reload and other threads can still be attempting (and succeeding) with authentication. Upon successful authentication we add to the cache.

We end up in situation where we can concurrently (across different threads) call Cache#put(k,v) (T#1) and Cache#invalidateAll() (T#20) on the same cache instance at the same time.  If you use a frequent rotation strategy with frequent calls that use the JWT realm can increase the likelyhood of those two method happening concurrently. These are the two calls that end up deadlocking each other (why is due to the cache implementation bug). If only those two threads deadlocked themselves then no one would probably notice since you are just down two threads. However, in the JWT realm you will notice because we attempt to read from the cache as an optimization for authentication. Since the two threads are deadlocked inside the cache implementation (the deadlocked threads hold the write locks) we can not read from the cache and end up blocking on the transport threads (where authentication is happening). At this point it is just a matter of (short) time before the entire cluster is useless since all the (http) transport threads will blocked due to the 2 deadlocked threads. 

It appears that the original author tried to protect against this situation by protected the put(k,v) and invalidateAll() by wrapping both of the calls in `try (ReleasableLock ignored = jwtCacheHelper.acquireUpdateLock()) { ` which on the surface appears to make put(k,v) and invalidateAll() mutually exclusive. However, if you look at the implementation of `acquireUpdateLock()` it is backed by *read* lock, which is not mutually exclusive to other read locks. The naming here is quite misleading. Using the read lock to provide mutually exclusive access is the first bug and fixing it will protect the JWT realm. Fixing it should be as easy as changing  `acquireUpdateLock()` to `acquireForIterator()` which is a *write* lock. There is probably a bigger discussion about the use of the `CacheIteratorHelper` without actually iterating ...but that can be addressed later. 

Fixing just the JWT realm side does not fix the potential for other deadlocks with a similar pattern of concurrent put(k,v) and invalidateAll() against the Cache. Ideally the cache would not require explicit locks for concurrent modifications and believe this a bug with the Cache implementation.

## Cache implementation bug

I will log another issue, but for completeness will include here too.

Looking at the stack trace, the relevant threads are T#1 which is trying to do the Cache#put(k,v) , and T#20 which is trying to do the Cache#invalidateAll(). T#12 (and 16 other threads) are blocked due to the deadlock between T#1 and T#20.  In summary here is the relevant work the cache does for each workflow: 
Line numbers are relevant to commit 2461cb152369d6ef92ce0203d97f70596a1e4ecb (HEAD of main at time of investigation). 


T#20 invalidateAll()
```
- locks segement write lock on 526 (reentrent r/w lock)  
- locks lru lock on 529 (reentrant lock)
- unlocks lru lock on 542
- unlocks segment write lock on 546
```

T#1 put(k,v)
```
- locks lru lock on 464 (reentrant lock)
- (re)locks lru lock on 739
- (evict/evictEntry/remove) locks segment write lock on 293 (reentrent r/w lock) 
- (evict/evictEntry/remove) unlocks segment write lock on 310
- (re)unlocks lru lock on 748
- unlocks lru lock on 471
```
To put it another way:
```
T#1  put(k,v)        -> lock lru -> lock swl -> unlock swl -> unlock lru
T#20 invalidateAll() -> lock swl -> lock lru -> unlock lru -> unlock swl
```
So the sequence of events for a dead lock must be 

1) T#1 grabs the lru lock
2) T#20 grabs segment write lock
3) T#1 tries to grab the segment write lock but is blocked by T#20
4) T#20 tries to grab the lru lock but is blocked by T#1

or some variant of that. The important bit is that put(k,v) and invalidateAll() must be happening at the same time across different threads and the attempts to grab the different 2 locks must be intertwined between the 2 different threads. 

Fixing this could be challenging and I have not explored if there other potential cases this deadlock could occur.

**Comment by ywangd** [comment]
This is a good find. Thanks a lot for the detailed analysis!

> It appears that the original author tried to protect against this situation by protected the put(k,v) and invalidateAll() by wrapping both of the calls in try (ReleasableLock ignored = jwtCacheHelper.acquireUpdateLock()) {  which on the surface appears to make put(k,v) and invalidateAll() mutually exclusive.

I think the use of [CacheHelper](url) was not trying to make `put` and `invalidateAll` mutually exclusive. It was itself a mis-use for a scenario that it does not apply. We (at least myself) were simply not aware of the fact `put` and `invalidateAll` can deadlock each other. By briefly reading the code, it seems `computeIfAbsent` can have the same problem as well. A fix for the `Cache` class would be of great help.
